File: sdks/rust/src/untrace.rs
//! Main Untrace SDK implementation

use crate::client::UntraceClient;
use crate::config::Config;
use crate::context::UntraceContext;
use crate::error::{UntraceError, UntraceResult};
use crate::instrumentation::Instrumentation;
use crate::metrics::UntraceMetrics;
use crate::provider::{ProviderRegistry, register_default_providers};
use crate::tracer::UntraceTracer;
use opentelemetry::global;
use opentelemetry::trace::TracerProvider;
use opentelemetry::KeyValue;
use opentelemetry_semantic_conventions::resource::SERVICE_NAME;
use std::sync::{Arc, OnceLock};
use tracing::{debug, info, warn};

/// Global Untrace instance
static GLOBAL_INSTANCE: OnceLock<Untrace> = OnceLock::new();

/// Main Untrace SDK struct
#[derive(Debug)]
pub struct Untrace {
    client: UntraceClient,
    instrumentation: Instrumentation,
    provider_registry: ProviderRegistry,
    config: Arc<Config>,
}

impl Untrace {
    /// Initialize the Untrace SDK
    pub async fn init(config: Config) -> UntraceResult<Self> {
        // Validate configuration
        config.validate()?;

        // Set up logging if debug is enabled
        if config.debug {
            tracing_subscriber::fmt()
                .with_env_filter("untrace=debug")
                .init();
            info!("Untrace SDK initialized with debug logging enabled");
        }

        // Create resource
        let mut resource_attributes = vec![
            KeyValue::new(SERVICE_NAME.to_string(), config.service_name.clone()),
            KeyValue::new("service.version", config.service_version.clone()),
            KeyValue::new("service.environment", config.environment.clone()),
        ];

        // Add custom resource attributes
        for (key, value) in &config.resource_attributes {
            resource_attributes.push(KeyValue::new(key.clone(), value.clone()));
        }

        // Set up OpenTelemetry
        let tracer = global::tracer("untrace-sdk");
        let untrace_tracer = Arc::new(UntraceTracer::new(tracer));

        // Create metrics
        let meter = global::meter("untrace-sdk");
        let metrics = Arc::new(UntraceMetrics::new(meter));

        // Create context
        let context = Arc::new(UntraceContext::new());

        // Create client
        let client = UntraceClient::new(untrace_tracer, metrics, context);

        // Create instrumentation
        let mut instrumentation = Instrumentation::new(Default::default());

        // Create provider registry
        let mut provider_registry = ProviderRegistry::new();
        register_default_providers(&mut provider_registry);

        // Enable instrumentation if not disabled
        if !config.disable_auto_instrumentation {
            instrumentation.enable()?;
        }

        let untrace = Self {
            client,
            instrumentation,
            provider_registry,
            config: Arc::new(config),
        };

        // Set global instance
        GLOBAL_INSTANCE.set(untrace.clone()).map_err(|_| {
            UntraceError::initialization("Failed to set global Untrace instance")
        })?;

        info!("Untrace SDK initialized successfully");
        Ok(untrace)
    }

    /// Initialize from environment variables
    pub async fn init_from_env() -> UntraceResult<Self> {
        let config = Config::from_env()?;
        Self::init(config).await
    }

    /// Get the global instance
    pub fn get_instance() -> Option<Self> {
        GLOBAL_INSTANCE.get().cloned()
    }

    /// Get the client
    pub fn client(&self) -> &UntraceClient {
        &self.client
    }

    /// Get the tracer
    pub fn tracer(&self) -> &UntraceTracer {
        self.client.tracer()
    }

    /// Get the metrics
    pub fn metrics(&self) -> &UntraceMetrics {
        self.client.metrics()
    }

    /// Get the context
    pub fn context(&self) -> &UntraceContext {
        self.client.context()
    }

    /// Get the instrumentation
    pub fn instrumentation(&self) -> &Instrumentation {
        &self.instrumentation
    }

    /// Get the provider registry
    pub fn provider_registry(&self) -> &ProviderRegistry {
        &self.provider_registry
    }

    /// Get the configuration
    pub fn config(&self) -> &Config {
        &self.config
    }

    /// Shutdown the SDK
    pub async fn shutdown(&self) -> UntraceResult<()> {
        info!("Shutting down Untrace SDK");

        // Disable instrumentation
        if self.instrumentation.is_enabled() {
            // Note: We can't mutate self here, so we'll just log
            warn!("Instrumentation is still enabled during shutdown");
        }

        // Shutdown client
        self.client.shutdown().await?;

        debug!("Untrace SDK shutdown complete");
        Ok(())
    }

    /// Flush any pending data
    pub async fn flush(&self) -> UntraceResult<()> {
        self.client.flush().await
    }
}

impl Clone for Untrace {
    fn clone(&self) -> Self {
        Self {
            client: self.client.clone(),
            instrumentation: self.instrumentation.clone(),
            provider_registry: self.provider_registry.clone(),
            config: Arc::clone(&self.config),
        }
    }
}

File: sdks/js/src/untrace.ts
import { DiagConsoleLogger, DiagLogLevel, diag } from '@opentelemetry/api';
import { registerInstrumentations } from '@opentelemetry/instrumentation';
import { resourceFromAttributes } from '@opentelemetry/resources';
import { BatchSpanProcessor } from '@opentelemetry/sdk-trace-base';
import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
import {
  ATTR_SERVICE_NAME,
  ATTR_SERVICE_VERSION,
} from '@opentelemetry/semantic-conventions';
import { UntraceContext } from './context';
import { UntraceExporter } from './exporter';
import { UntraceMetricsImpl } from './metrics';
import { getProviderInstrumentation } from './providers';
import { UntraceTracer } from './tracer';
import type { ProviderInstrumentation, UntraceConfig } from './types';

// Global state management
const globalState = {
  instance: null as Untrace | null,
  isDiagSet: false,
  isInitialized: false,
  isRegistered: false,
};

/**
 * Main Untrace SDK class
 */
export class Untrace {
  private provider: NodeTracerProvider;
  private config: Required<UntraceConfig>;
  private tracer: UntraceTracer;
  private metrics: UntraceMetricsImpl;
  private context: UntraceContext;
  private instrumentations: ProviderInstrumentation[] = [];
  private httpInstrumentation?: any; // HTTPLLMInstrumentation instance
  private llmInstrumentationManager?: any; // LLMInstrumentationManager instance

  constructor(config: UntraceConfig) {
    // Set up config with defaults
    this.config = {
      apiKey: config.apiKey,
      baseUrl: config.baseUrl || 'https://untrace.dev',
      captureBody: config.captureBody !== false,
      captureErrors: config.captureErrors !== false,
      debug: config.debug || false,
      disableAutoInstrumentation: config.disableAutoInstrumentation || false,
      exportIntervalMs: config.exportIntervalMs || 5000,
      headers: config.headers || {},
      maxBatchSize: config.maxBatchSize || 512,
      providers: config.providers || ['all'],
      resourceAttributes: config.resourceAttributes || {},
      samplingRate: config.samplingRate || 1.0,
      spanProcessors: config.spanProcessors || [],
      version: config.version || '0.0.0',
    };

    // Create resource with service information (simplified like test-sdk.js)
    if (this.config.debug && !globalState.isDiagSet) {
      diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.DEBUG);
      globalState.isDiagSet = true;
    }
    const resource = resourceFromAttributes({
      [ATTR_SERVICE_NAME]: 'untrace-app',
      [ATTR_SERVICE_VERSION]: this.config.version,
      ...this.config.resourceAttributes,
    });

    // Create exporter
    const exporter = new UntraceExporter(this.config);

    // Create batch span processor (simplified like test-sdk.js)
    const batchSpanProcessor = new BatchSpanProcessor(exporter, {
      maxExportBatchSize: this.config.maxBatchSize,
      maxQueueSize: Math.max(this.config.maxBatchSize * 2, 10), // Ensure queue is at least 2x batch size or 10
      scheduledDelayMillis: this.config.exportIntervalMs,
    });

    // Create tracer provider
    this.provider = new NodeTracerProvider({
      resource,
      spanProcessors: [batchSpanProcessor],
    });

    // Register provider
    this.provider.register();

    // Initialize components
    this.tracer = new UntraceTracer(
      this.provider.getTracer('untrace-app', this.config.version),
    );
    this.metrics = new UntraceMetricsImpl();
    this.context = new UntraceContext();

    // Auto-instrument if not disabled
    if (!this.config.disableAutoInstrumentation) {
      this.autoInstrument();
    }

    // Store instance
    globalState.instance = this;
  }

  /**
   * Auto-instrument LLM SDKs and providers using the comprehensive matrix approach
   *
   * This approach uses a matrix of LLM SDKs and providers to determine the best
   * instrumentation strategy for each combination, ensuring comprehensive coverage
   * regardless of custom configurations or base URL overrides.
   */
  private autoInstrument(): void {
    // Import LLM instrumentation manager dynamically
    import('./instrumentation/manager').then(({ LLMInstrumentationManager }) => {
      const llmManager = new LLMInstrumentationManager({
        enabled: true,
      });

      // Enable all supported instrumentations
      llmManager.enableAllInstrumentations().then(() => {
        // Store reference for cleanup
        this.llmInstrumentationManager = llmManager;

        if (this.config.debug) {
          const status = llmManager.getInstrumentationStatus();
          const supported = llmManager.getSupportedCombinations();
          console.log('[Untrace] LLM instrumentation enabled:', {
            active: Object.keys(status).length,
            supported: supported.length,
            status
          });
        }
      }).catch((error) => {
        if (this.config.debug) {
          console.warn('[Untrace] Failed to enable some LLM instrumentations:', error);
        }
      });
    }).catch((error) => {
      if (this.config.debug) {
        console.warn('[Untrace] Failed to load LLM instrumentation manager:', error);
      }
    });
  }

  /**
   * Get the tracer instance
   */
  getTracer(): UntraceTracer {
    return this.tracer;
  }

  /**
   * Get the metrics instance
   */
  getMetrics(): UntraceMetricsImpl {
    return this.metrics;
  }

  /**
   * Get the context instance
   */
  getContext(): UntraceContext {
    return this.context;
  }

  /**
   * Manually instrument a provider
   */
  instrument<T>(providerName: string, module: T): T {
    const instrumentation = getProviderInstrumentation(providerName);
    if (!instrumentation) {
      throw new Error(`Unknown provider: ${providerName}`);
    }

    if (!instrumentation.canInstrument(module)) {
      throw new Error(
        `Module is not instrumentable by ${providerName} provider`,
      );
    }

    instrumentation.initialize(this.config, this.tracer.getTracer());
    return instrumentation.instrument(module);
  }

  /**
   * Flush all pending spans
   */
  async flush(): Promise<void> {
    if (this.config.debug) {
      console.log('[Untrace] Flushing spans...');
    }
    await this.provider.forceFlush();
    if (this.config.debug) {
      console.log('[Untrace] Flush completed');
    }
  }

  /**
   * Shutdown the SDK
   */
  async shutdown(): Promise<void> {
    if (this.config.debug) {
      console.log('[Untrace] Shutting down SDK...');
    }

    // Disable all instrumentations
    this.instrumentations.forEach((inst) => {
      try {
        inst.disable();
      } catch (_error) {
        // Ignore errors during shutdown
      }
    });

    // Disable HTTP instrumentation
    if (this.httpInstrumentation) {
      try {
        this.httpInstrumentation.disable();
      } catch (_error) {
        // Ignore errors during shutdown
      }
    }

    // Disable LLM instrumentation manager
    if (this.llmInstrumentationManager) {
      try {
        this.llmInstrumentationManager.disableAllInstrumentations();
      } catch (_error) {
        // Ignore errors during shutdown
      }
    }

    // Shutdown provider
    await this.provider.shutdown();

    // Clear instance and reset singleton state
    if (globalState.instance === this) {
      globalState.instance = null;
      globalState.isInitialized = false;
    }
    globalState.isRegistered = false;

    if (this.config.debug) {
      console.log('[Untrace] SDK shutdown complete');
    }
  }

  /**
   * Get the current Untrace instance
   */
  static getInstance(): Untrace | null {
    return globalState.instance;
  }
}

/**
 * Initialize the Untrace SDK (simplified)
 */
export function init(config: UntraceConfig): Untrace {
  if (globalState.isInitialized && globalState.instance) {
    if (config.debug) {
      console.log(
        '[Untrace] SDK already initialized. Returning existing instance.',
      );
    }
    return globalState.instance;
  }

  if (globalState.isInitialized) {
    throw new Error('Untrace SDK is already initialized but instance is null.');
  }

  globalState.isInitialized = true;
  globalState.instance = new Untrace(config);
  return globalState.instance;
}

/**
 * Get the current Untrace instance
 */
export function getUntrace(): Untrace {
  if (!globalState.instance) {
    throw new Error('Untrace SDK not initialized. Call init() first.');
  }
  return globalState.instance;
}
